src/runtime/tracing.cpp

```diff
@@ -183,16 +286,18 @@ WEAK trace_fn halide_set_custom_trace(trace_fn t) {
 
 WEAK void halide_set_trace_file(int fd) {
     halide_trace_file = fd;
-    halide_trace_file_initialized = true;
 }
 
 extern int errno;
 
 WEAK int halide_get_trace_file(void *user_context) {
-    // Prevent multiple threads both trying to initialize the trace
-    // file at the same time.
-    ScopedSpinLock lock(&halide_trace_file_lock);
-    if (!halide_trace_file_initialized) {
+    if (halide_trace_file >= 0) {
+        return halide_trace_file;
+    }
+    {
+        // Prevent multiple threads both trying to initialize the trace
+        // file at the same time.
+        ScopedSpinLock lock(&halide_trace_file_lock);
```

Comment:
You definitely have to check if another thread beat this one into the critical section before opening the file...